src/snowflake/snowpark/stored_procedure_profiler.py

```diff
@@ -0,0 +1,131 @@
+            registered modules.
+        Args:
+            stored_procedures: List of names of stored procedures.
+        """
+        sql_statement = (
+            f"alter session set python_profiler_modules='{','.join(stored_procedures)}'"
+        )
+        self._session.sql(sql_statement)._internal_collect_with_tag_no_telemetry()
+
+    def set_targeted_stage(self, stage: str):
+        """
+        Set targeted stage for profiler output.
+
+        Note:
+            The stage name must be a fully qualified name.
+
+        Args:
+            stage: String of fully qualified name of targeted stage
+        """
+        validate_object_name(stage)
+        if (
+            len(
+                self._session.sql(  # type: ignore
+                    f"show stages like '{stage}'"
+                )._internal_collect_with_tag_no_telemetry()
+            )
+            == 0
+            and len(
+                self._session.sql(  # type: ignore
+                    f"show stages like '{stage.split('.')[-1]}'"
+                )._internal_collect_with_tag_no_telemetry()
+            )
+            == 0
+        ):
+            self._session.sql(
+                f"create temp stage if not exists {stage} FILE_FORMAT = (RECORD_DELIMITER = NONE FIELD_DELIMITER = NONE )"
+            )._internal_collect_with_tag_no_telemetry()
+        sql_statement = f'alter session set PYTHON_PROFILER_TARGET_STAGE ="{stage}"'
+        self._session.sql(sql_statement)._internal_collect_with_tag_no_telemetry()
+
+    def set_active_profiler(self, active_profiler_type: Literal["LINE", "MEMORY"]):
+        """
+        Set active profiler.
+
+        Note:
+            Active profiler must be either 'LINE' or 'MEMORY' (case-sensitive). Active profiler is 'LINE' by default.
+        Args:
+            active_profiler_type: String that represent active_profiler, must be either 'LINE' or 'MEMORY'
+            (case-sensitive).
+
+        """
+        if active_profiler_type not in ["LINE", "MEMORY"]:
+            raise ValueError(
+                f"active_profiler expect 'LINE', 'MEMORY', got {active_profiler_type} instead"
+            )
+        sql_statement = (
+            f"alter session set ACTIVE_PYTHON_PROFILER = '{active_profiler_type}'"
+        )
+        self._session.sql(sql_statement)._internal_collect_with_tag_no_telemetry()
+        self._query_history = self._session.query_history(include_thread_id=True)
+
+    def disable(self) -> None:
+        """
+        Disable profiler.
+        """
+        self._session._conn.remove_query_listener(self._query_history)  # type: ignore
+        sql_statement = "alter session set ACTIVE_PYTHON_PROFILER = ''"
+        self._session.sql(sql_statement)._internal_collect_with_tag_no_telemetry()
+
+    @staticmethod
+    def _is_sp_call(query: str) -> bool:
+        return re.match(
+            STORED_PROCEDURE_CALL_PATTERN, query.strip(" "), re.DOTALL
```

Comment:
1. we can do the `strip()` for both once.
2. is this match case insensitive? if not, we can probably do `query = query.strip().upper()` as a common step above.